### PR TITLE
Change of approach to data storage. AC, A2C changed

### PR DESCRIPTION
**Pull request by kmakeev**
Hello!
Again, I took the liberty of making changes to the code without approval.
For optimization, I tried to change the data storage approach:
1. Data is stored in the form in which it is transmitted by the environment.
2. Conversion to float32 takes place before they are sent to the model or in the tf-function itself.
 I tried on AC, A2C Algorithms .. It seems to work and requires much less RAM.
I’ll test it at home on a more powerful PC on Atari Env.
As a suggestion, I have the following:
go completely to type float64;
try to perform data conversion for all algorithms only in tf-functions.

### CODE REVIEW

**Comment by StepNeverStop** [comment]
Hi, @kmakeev 

I've forgot why I write so many `.astype(np.float32)` before. But it seems I have encountered to a bug and must done this to fix it.

I'm not sure if accepted this PR whether the bug will occur or not, plz do more test on different gym enviornments if it's convenient for you.

thank you very much. Because I have been busy with other things recently, I will deal with this problem later

**Comment by StepNeverStop** [comment]
> Hello!
> Again, I took the liberty of making changes to the code without approval.
> For optimization, I tried to change the data storage approach:
> 
> 1. Data is stored in the form in which it is transmitted by the environment.
> 2. Conversion to float32 takes place before they are sent to the model or in the tf-function itself.
>     I tried on AC, A2C Algorithms .. It seems to work and requires much less RAM.
>    I’ll test it at home on a more powerful PC on Atari Env.
>    As a suggestion, I have the following:
>    go completely to type float64;
>    try to perform data conversion for all algorithms only in tf-functions.

Actually, I've encounter a gym environment which reward has type of `int`. That's why I write `reward.astype(np.float32)` in `gym_wrapper.py`.

I know what you mean, you want deal with the data type when extracting experiences from the replay buffer so that could save some memory. But it's a trade-off between training speed and memory usage here:
1. change data type when storing it. High training speed, much memory use(maybe)
2. change data type when extracting and using it. Low training speed(beceause one experience could be extracted many times, so there will be multiple conversions), less memory use(maybe).

And I still note that you added an normalized operation for visual input. Actually, I will add this function in `gym_wrapper.py` deal with both vector and visual input.

**Comment by kmakeev** [comment]
> And I still note that you added an normalized operation for visual input. Actually, I will add this  function in gym_wrapper.py deal with both vector and visual input.

You can clarify for a better understanding.

> 2.change data type when extracting and using it. Low training speed(beceause one experience could be extracted many times, so there will be multiple conversions), less memory use(maybe).

Can be easily checked by tests.

**Comment by StepNeverStop** [comment]
> > And I still note that you added an normalized operation for visual input. Actually, I will add this  function in gym_wrapper.py deal with both vector and visual input.
> 
> You can clarify for a better understanding.
> 
> > 2.change data type when extracting and using it. Low training speed(beceause one experience could be extracted many times, so there will be multiple conversions), less memory use(maybe).
> 
> Can be easily checked by tests.

I noted you added `features = features / 255` in `tf2nn.py`, which is a normalization operation for visual input. Actually, I would like to implement observation normalizetion in `gym_wrapper.py` layer. And whether change [210, 160, 3] to [84, 84, 3] or grayscale is on the way !

**Comment by kmakeev** [comment]
OK! Now I understand.
But, all these cases also require normalization, normalization is needed for any visual inputs.
In my opinion, it is better to do this in the tf function, the calculation will be on the GPU.
In gym_wrapper.py I did not find calculations on tensors ...

**Comment by StepNeverStop** [comment]
@kmakeev 

That's a good question... I‘ll think it twice.

And, does `tf.cast` run on GPU? I dont know that, do you know?

**Comment by kmakeev** [comment]
Must.
I will check it.

**Comment by StepNeverStop** [comment]
If the `np` operation is mixed in `tf`, i.e. tf.max(np.log(x)) will it affect the speed of GPU? Do you know about this?

**Comment by kmakeev** [comment]
> If the np operation is mixed in tf, i.e. tf.max(np.log(x)), will it affect the speed of GPU? Do you know about this?
I checked on this code.
from __future__ import absolute_import, division, print_function, unicode_literals
import tensorflow as tf
import numpy as np
import datetime


def normalizing_(a):
    b = tf.cast(a, dtype=tf.float64)
    b = b / 255
    for i in range(10):
        print("Executed numpy", np.amax(np.log(b)))


@tf.function
def normalizing(a):
    b = tf.cast(a, dtype=tf.float64)
    b = b / 255
    for i in range(10):
        tf.print("Executed tf - ", tf.reduce_max(tf.math.log(b)))


array = np.random.randint(0, 255, size=10**8, dtype = np.uint8)

print(datetime.datetime.now())
with tf.device('/CPU:0'):
    result = normalizing(array)

print(datetime.datetime.now())
with tf.device('/GPU:0'):
    result = normalizing(array)

print(datetime.datetime.now())
result = normalizing_(array)
print(datetime.datetime.now())

2019-11-20 21:22:45.303259
Executed tf -  -0.00392927813988955
Executed tf -  -0.00392927813988955
2019-11-20 21:22:46.271348
Executed tf -  -0.00392927813988955
Executed tf -  -0.00392927813988955
Executed tf -  -0.00392927813988955
Executed tf -  -0.00392927813988955
Executed tf -  -0.00392927813988955
Executed tf -  -0.00392927813988955
Executed tf -  -0.00392927813988955
Executed tf -  -0.00392927813988955
2019-11-20 21:22:46.395311
Executed tf -  -0.00392927813988955
Executed tf -  -0.00392927813988955
Executed tf -  -0.00392927813988955
Executed tf -  -0.00392927813988955
Executed tf -  -0.00392927813988955
Executed tf -  -0.00392927813988955
Executed tf -  -0.00392927813988955
Executed tf -  -0.00392927813988955
Executed tf -  -0.00392927813988955
Executed tf -  -0.00392927813988955
Executed numpy -0.00392927813988955
Executed numpy -0.00392927813988955
Executed numpy -0.00392927813988955
Executed numpy -0.00392927813988955
Executed numpy -0.00392927813988955
Executed numpy -0.00392927813988955
Executed numpy -0.00392927813988955
Executed numpy -0.00392927813988955
Executed numpy -0.00392927813988955
Executed numpy -0.00392927813988955
2019-11-20 21:22:51.949725

Process finished with exit code 0


Everything works, much faster in tf-functions. on the GPU will be more noticeable with more complex calculations

**Comment by kmakeev** [comment]
I have already tested AC, A2C algorithms with my code. While everything works, there are no errors.
GPU utilization reaches 100%, especially on the AC algorithm. Let's see what he can achieve.

Снимок экрана от 2019-11-20 23-28-40

**Comment by StepNeverStop** [comment]
> > If the np operation is mixed in tf, i.e. tf.max(np.log(x)), will it affect the speed of GPU? Do you know about this?
> 
> I checked on this code.
> from **future** import absolute_import, division, print_function, unicode_literals
> import tensorflow as tf
> import numpy as np
> import datetime
> 
> def normalizing_(a):
> b = tf.cast(a, dtype=tf.float64)
> b = b / 255
> for i in range(10):
> print("Executed numpy", np.amax(np.log(b)))
> 
> @tf.function
> def normalizing(a):
> b = tf.cast(a, dtype=tf.float64)
> b = b / 255
> for i in range(10):
> tf.print("Executed tf - ", tf.reduce_max(tf.math.log(b)))
> 
> array = np.random.randint(0, 255, size=10**8, dtype = np.uint8)
> 
> print(datetime.datetime.now())
> with tf.device('/CPU:0'):
> result = normalizing(array)
> 
> print(datetime.datetime.now())
> with tf.device('/GPU:0'):
> result = normalizing(array)
> 
> print(datetime.datetime.now())
> result = normalizing_(array)
> print(datetime.datetime.now())
> 
> 2019-11-20 21:22:45.303259
> Executed tf - -0.00392927813988955
> Executed tf - -0.00392927813988955
> 2019-11-20 21:22:46.271348
> Executed tf - -0.00392927813988955
> Executed tf - -0.00392927813988955
> Executed tf - -0.00392927813988955
> Executed tf - -0.00392927813988955
> Executed tf - -0.00392927813988955
> Executed tf - -0.00392927813988955
> Executed tf - -0.00392927813988955
> Executed tf - -0.00392927813988955
> 2019-11-20 21:22:46.395311
> Executed tf - -0.00392927813988955
> Executed tf - -0.00392927813988955
> Executed tf - -0.00392927813988955
> Executed tf - -0.00392927813988955
> Executed tf - -0.00392927813988955
> Executed tf - -0.00392927813988955
> Executed tf - -0.00392927813988955
> Executed tf - -0.00392927813988955
> Executed tf - -0.00392927813988955
> Executed tf - -0.00392927813988955
> Executed numpy -0.00392927813988955
> Executed numpy -0.00392927813988955
> Executed numpy -0.00392927813988955
> Executed numpy -0.00392927813988955
> Executed numpy -0.00392927813988955
> Executed numpy -0.00392927813988955
> Executed numpy -0.00392927813988955
> Executed numpy -0.00392927813988955
> Executed numpy -0.00392927813988955
> Executed numpy -0.00392927813988955
> 2019-11-20 21:22:51.949725
> 
> Process finished with exit code 0
> 
> Everything works, much faster in tf-functions. on the GPU will be more noticeable with more complex calculations

Hi, @kmakeev 

from your result , I didn't see any difference between using `numpy` or `tensorflow` operation.

Now, I've changed the time when changing the type of data, you can see the [lattest commit](https://github.com/StepNeverStop/RLs/commit/7072b580ef2d5c0c0f0ce46742c6847e6dee9008) for detail.

**Comment by kmakeev** [comment]
Hi!
> from your result , I didn't see any difference between using numpy or tensorflow operation.
Now, I've changed the time when changing the type of data, you can see the lattest commit for detail.
The difference is clearly visible in the visual execution of the code.
A time stamp was printed between function calls. From it the difference is visible.
This is confirmed by other sources.

PR can be closed, I will understand further in your master branch.
https://databricks.com/tensorflow/using-a-gpu
PR can be closed, I will understand further in your master branch.
